plugins/chain_plugin/chain_plugin.cpp

```diff
@@ -604,6 +610,10 @@ void chain_plugin::plugin_initialize(const variables_map& options) {
          throw;
       }
 
+      eosio::chain::chain_config::deep_mind_enabled = options.at( "deep-mind" ).as<bool>();
```

Comment:
Instead of this I suggest the following:
- Create an `fc::logger _deep_mind_log` in `chain_plugin.cpp` in a similar way that `_trx_trace_log` was created in `producer_plugin.cpp` (including adding a `handle_sighup` to `chain_plugin`). 
- Add a `fc::logger* deep_mind_log = nullptr;` member variable to `controller_impl`. 
- Add an inline member function `fc::logger* controller_impl::get_deep_mind_logger()const` which simply returns `deep_mind_log`.
- Add a member function `fc::logger* controller::get_deep_mind_logger()const` which simply returns `my->get_deep_mind_logger()`.
- Add a member function `void controller::enable_deep_mind( fc::logger* log )` which sets the `my->deep_mind_log` pointer to the provided value if it is currently `nullptr` (you could assert if wasn't `nullptr`).
- Finally, within `chain_plugin::plugin_initialize`, if and only if `options.at( "deep-mind" ).as<bool>()` is true, call `my->chain->enable_deep_mind( &_deep_mind_log )`. 

Then everywhere you current do:
```cpp
if (eosio::chain::chain_config::deep_mind_enabled) {
```
could be replaced with:
```cpp
if (auto dm_log = get_deep_mind_logger()) {
```
or:
```cpp
if (auto dm_log = control.get_deep_mind_logger()) {
```
or something similar depending on where it is at (`controller`, `apply_context`, `transaction_context`, etc.). You will probably need to add a reference to `controller` into `resource_limits_manager` though. Also instead of including the one log within `protocol_feature_manager.cpp`, I would instead modify `protocol_feature_manager::activate_feature` to return `protocol_feature_manager::const_iterator` pointing to the newly activated protocol feature and allow the caller in `controller_impl::start_block` to do the logging.

Then within the if statement you can log by doing something like the following:
```cpp
   fc_dlog(*dm_log, "message", ...);
```

Reply:
All applied. 

I decided on not implementing the suggested `activate_feature` change. The `activate_feature` is used on other code paths than just `start_block`, so it seemed more safe to stay in there so the log print is output for all paths.

Reply:
Also, instead of using`fc_dlog(*dm_log, "message", ...);`, I adjusted the `dmlog` macro to use the logger and write directly, making it as lean as possible, and more readable (IMO).